# Dataset 1: EU population aged 55+


## Phase 1: data familiarisation

## Data source
- Eurostat: Population on 1 January by age group and sex (demo_pjangroup)
- Link: https://ec.europa.eu/eurostat/databrowser/view/demo_pjangroup__custom_21849453/default/table
- The original dataset was filtered for the 2011-2025 timeframe, all age categories as of 55+ and the EU total, and the EU27 countries and the total

In [1]:
import os
os.chdir(os.path.dirname(os.path.abspath('__file__')))

In [2]:
import pandas as pd
df = pd.read_csv('demo_pjangroup__eu27_as_of_age_55.tsv', sep='\t')

In [3]:
df

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,"A,NR,F,TOTAL,AT",4296293,4309977,4328238,4352447,4384529,4427918,4460424,4483749,4501742,4522292,4535712,4553444,4619957,4643918,4664333
1,"A,NR,F,TOTAL,BE",5598920 b,5635601 b,5665118,5687048,5713206,5741853,5762455,5784371,5810693,5841215,5854293,5883978,5946649,5981093,6012626
2,"A,NR,F,TOTAL,BG",3780505,3760457,3701945,3662824,3621883,3578998,3537549,3494184,3450251,3405861,3389807,3364303,3348207,3347783,3342220
3,"A,NR,F,TOTAL,CY",430993,443018,443058,441537,441559,442623,445880,450423,455583,459191,466924,476513,483955,490771,498765
4,"A,NR,F,TOTAL,CZ",5339500,5347235,5351776,5350039,5361348,5367513,5378133,5390264,5405606,5421943,5323934 b,5332932,5519006,5557945,5554921
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
667,"A,NR,T,Y_GE85,PT",225171,252838,262367,274828,287550,298622,310077,322329,332277,343512,352726,362327,368507,379366,388556
668,"A,NR,T,Y_GE85,RO",261090,264308,282860,305065,323182,341394,357254,373881,387272,403733 e,411247 e,360798 e,370171 e,389826 e,408898 e
669,"A,NR,T,Y_GE85,SE",251217,253327,252968,255029,256484,258471,259864,260549,260606,263256,262690,268623,274016,282309,293545
670,"A,NR,T,Y_GE85,SI",34098,36369,38735,41175,43382,46171,48288,50395,52276,54136,54575,56004,57136,58866,60718


In [4]:
# Checking first 5 rows

df.head(10)

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,"A,NR,F,TOTAL,AT",4296293,4309977,4328238,4352447,4384529,4427918,4460424,4483749,4501742,4522292,4535712,4553444,4619957,4643918,4664333
1,"A,NR,F,TOTAL,BE",5598920 b,5635601 b,5665118,5687048,5713206,5741853,5762455,5784371,5810693,5841215,5854293,5883978,5946649,5981093,6012626
2,"A,NR,F,TOTAL,BG",3780505,3760457,3701945,3662824,3621883,3578998,3537549,3494184,3450251,3405861,3389807,3364303,3348207,3347783,3342220
3,"A,NR,F,TOTAL,CY",430993,443018,443058,441537,441559,442623,445880,450423,455583,459191,466924,476513,483955,490771,498765
4,"A,NR,F,TOTAL,CZ",5339500,5347235,5351776,5350039,5361348,5367513,5378133,5390264,5405606,5421943,5323934 b,5332932,5519006,5557945,5554921
5,"A,NR,F,TOTAL,DE_TOT",41097465 b,41097953,41142770,41210540,41362080,41661561,41824535,41948786,42052522,42129098,42128512,42170339,42198796 b,42294114,42335439
6,"A,NR,F,TOTAL,DK",2804046,2813740,2823776,2834956,2848701,2869364,2888591,2904717,2917008,2925845,2935188,2950505,2983655,2997558,3013749
7,"A,NR,F,TOTAL,EE",709960,707079,704004,700900,698882,699236,698097,698049,699185,699643,699153,698381,718518,723074,720242
8,"A,NR,F,TOTAL,EL",:,5655130,5624086,5594831,5565739,5529331,5508324,5482937,5457214,5431727,5398117,5346259,5315001,5292182,5278241
9,"A,NR,F,TOTAL,ES",23617694,23719207,23699424,23621034,23603001,23612439,23666286,23755470,23902168,24107382,24152187,24198046,24519768,24792824,25037928


In [5]:
# Checking last 5 rows

df.tail()

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
667,"A,NR,T,Y_GE85,PT",225171,252838,262367,274828,287550,298622,310077,322329,332277,343512,352726,362327,368507,379366,388556
668,"A,NR,T,Y_GE85,RO",261090,264308,282860,305065,323182,341394,357254,373881,387272,403733 e,411247 e,360798 e,370171 e,389826 e,408898 e
669,"A,NR,T,Y_GE85,SE",251217,253327,252968,255029,256484,258471,259864,260549,260606,263256,262690,268623,274016,282309,293545
670,"A,NR,T,Y_GE85,SI",34098,36369,38735,41175,43382,46171,48288,50395,52276,54136,54575,56004,57136,58866,60718
671,"A,NR,T,Y_GE85,SK",59798,62421,64752,67583,70733,73446,76447,79065,81207,83576,85081,78985,79255,82154,85582


No footer rows

In [6]:
# Counting rows and columns

df.shape
print(f"Dataset dimensions: {df.shape[0]} rows x {df.shape[1]} columns (excluding header)")

Dataset dimensions: 672 rows x 16 columns (excluding header)


In [7]:
# Checking column names, data types, non-null counts
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 672 entries, 0 to 671
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   freq,unit,sex,age,geo\TIME_PERIOD  672 non-null    object
 1   2011                               672 non-null    object
 2   2012                               672 non-null    object
 3   2013                               672 non-null    object
 4   2014                               672 non-null    object
 5   2015                               672 non-null    object
 6   2016                               672 non-null    object
 7   2017                               672 non-null    object
 8   2018                               672 non-null    object
 9   2019                               672 non-null    object
 10  2020                               672 non-null    object
 11  2021                               672 non-null    object
 12  2022    

## Findings from df.info()
- All columns read as object type due to Eurostat flags mixed into values
- No nulls detected at this stage but expected after value/flag separation
- First column contains combined dimensions - needs splitting

In [8]:
# Checking basic statistics on numeric columns
df.describe()

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
count,672,672,672,672,672,672,672,672,672,672,672,672,672,672,672,672
unique,672,649,672,669,670,670,669,670,670,670,670,671,672,671,672,671
top,"A,NR,F,TOTAL,AT",:,4309977,:,:,:,:,:,:,:,:,61052,4553444,69235,4643918,5695
freq,1,24,1,3,3,3,3,3,3,3,3,2,1,2,1,2


The':' is the Eurostat code for data not available (missing value).,There are missing values for 2011 and for 2013-2020.

In [9]:
# Checkig the exact column names
df.columns.tolist()

['freq,unit,sex,age,geo\\TIME_PERIOD',
 '2011 ',
 '2012 ',
 '2013 ',
 '2014 ',
 '2015 ',
 '2016 ',
 '2017 ',
 '2018 ',
 '2019 ',
 '2020 ',
 '2021 ',
 '2022 ',
 '2023 ',
 '2024 ',
 '2025 ']

Column names need stripping before use.

In [10]:
print(df['2011 '].unique())

['4296293 ' '5598920 b' '3780505 ' '430993 ' '5339500 ' '41097465 b'
 '2804046 ' '709960 ' ': ' '23617694 ' '225708502 b' '2736860 '
 '33515819 ' '2221105 ' '5241821 ' '2301455 ' '30933668 ' '1645365 '
 '257221 ' '1126666 ' '208584 ' '8412317 ' '19633018 ' '5519178 '
 '10366110 ' '4725326 ' '1035626 ' '2767273 ' '255710 ' '352844 '
 '273320 ' '24134 ' '386766 ' '2733178 b' '174884 ' '48500 ' '1342276 '
 '14779215 ' '193971 ' '2149654 ' '157418 ' '404366 ' '120839 ' '1907566 '
 '102162 ' '14990 ' '74277 ' '14428 ' '543295 ' '1501318 ' '353767 '
 '762308 ' '284742 ' '74205 ' '200157 ' '244484 ' '330191 ' '291796 '
 '22948 ' '391620 ' '2341656 b' '183462 ' '45650 ' '1257854 ' '13860143 '
 '204664 ' '2119617 ' '142264 ' '342201 ' '107692 ' '1975960 ' '94094 '
 '12828 ' '69621 ' '15586 ' '550206 ' '1261267 ' '330326 ' '671112 '
 '308051 ' '62681 ' '163627 ' '224142 ' '249981 ' '232088 ' '17430 '
 '302620 ' '2253848 b' '162422 ' '36630 ' '1102779 ' '11096775 ' '147454 '
 '1408822 ' '113884 '

Both column names and values contain leading/trailing whitespace. 

In [11]:
# Strip whitespace from all values
df = df.apply(lambda x: x.str.strip() if x.dtype == 'object' else x)

In [12]:
df.columns = df.columns.str.strip()

In [13]:
print(df['2011'].unique())

['4296293' '5598920 b' '3780505' '430993' '5339500' '41097465 b' '2804046'
 '709960' ':' '23617694' '225708502 b' '2736860' '33515819' '2221105'
 '5241821' '2301455' '30933668' '1645365' '257221' '1126666' '208584'
 '8412317' '19633018' '5519178' '10366110' '4725326' '1035626' '2767273'
 '255710' '352844' '273320' '24134' '386766' '2733178 b' '174884' '48500'
 '1342276' '14779215' '193971' '2149654' '157418' '404366' '120839'
 '1907566' '102162' '14990' '74277' '14428' '543295' '1501318' '353767'
 '762308' '284742' '74205' '200157' '244484' '330191' '291796' '22948'
 '391620' '2341656 b' '183462' '45650' '1257854' '13860143' '204664'
 '2119617' '142264' '342201' '107692' '1975960' '94094' '12828' '69621'
 '15586' '550206' '1261267' '330326' '671112' '308051' '62681' '163627'
 '224142' '249981' '232088' '17430' '302620' '2253848 b' '162422' '36630'
 '1102779' '11096775' '147454' '1408822' '113884' '298838' '85286'
 '1606421' '90956' '10109' '65686' '11089' '399835' '765627' '293862'
 '4

In [14]:
df[df['2011'] == ':']

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
8,"A,NR,F,TOTAL,EL",:,5655130,5624086,5594831,5565739,5529331,5508324,5482937,5457214,5431727,5398117,5346259,5315001,5292182,5278241
36,"A,NR,F,Y55-59,EL",:,349869,354746,362285,369381,374164,375236,376798,376267,374735,374926,381477,391000,400602,405555
64,"A,NR,F,Y60-64,EL",:,331100,328276,326881,330565,335712,341115,345971,352997,359840,364711,365629,367794,368395,368064
92,"A,NR,F,Y65-69,EL",:,289632,303278,314948,319216,319472,315985,312672,313183,319155,324410,329632,335390,342938,350321
120,"A,NR,F,Y70-74,EL",:,288609,274233,261137,256632,258684,271868,287431,299659,301085,300942,295657,293483,296437,304318
148,"A,NR,F,Y75-79,EL",:,276967,278930,279601,276544,270900,259688,246544,233533,230649,231826,245989,261844,273135,273855
176,"A,NR,F,Y80-84,EL",:,209131,215294,219752,225891,227439,230388,231594,232306,228369,224360,213124,201324,190586,190281
204,"A,NR,F,Y_GE85,EL",:,151304,160335,174580,187706,200931,212997,221337,230982,238946,242613,242502,243584,248609,252579
232,"A,NR,M,TOTAL,EL",:,5417595,5355920,5299463,5252302,5206084,5199203,5184399,5177999,5180864,5157227,5115368,5086867,5083582,5094094
260,"A,NR,M,Y55-59,EL",:,330735,332504,336789,341261,343268,344146,345849,346527,347180,349671,357840,367872,377832,383635


In [15]:
df[df['2013'] == ':']

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
206,"A,NR,F,Y_GE85,EU27_2020",6717973,6996707,:,:,:,:,:,:,:,:,8708349 bep,8806127 ep,8880717 bep,9075199 ep,9345128 ep
430,"A,NR,M,Y_GE85,EU27_2020",2755816,2919828,:,:,:,:,:,:,:,:,4288087 bep,4365444 ep,4434772 bep,4604581 ep,4822775 ep
654,"A,NR,T,Y_GE85,EU27_2020",9473789,9916535,:,:,:,:,:,:,:,:,12996436 bep,13171571 ep,13315489 bep,13679780 ep,14167903 ep


In [16]:
df[df['2020'] == ':']

,"freq,unit,sex,age,geo\TIME_PERIOD",2011,2012,2013,2014,2015,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
206,"A,NR,F,Y_GE85,EU27_2020",6717973,6996707,:,:,:,:,:,:,:,:,8708349 bep,8806127 ep,8880717 bep,9075199 ep,9345128 ep
430,"A,NR,M,Y_GE85,EU27_2020",2755816,2919828,:,:,:,:,:,:,:,:,4288087 bep,4365444 ep,4434772 bep,4604581 ep,4822775 ep
654,"A,NR,T,Y_GE85,EU27_2020",9473789,9916535,:,:,:,:,:,:,:,:,12996436 bep,13171571 ep,13315489 bep,13679780 ep,14167903 ep


In [17]:
# Check EU27_2020 Y_GE85 missing values across all years
key = df.columns[0]
eu27_yge85 = df[df[key].str.contains('EU27_2020') & df[key].str.contains('Y_GE85')]
year_cols = [str(y) for y in range(2011 , 2026 )]
print(eu27_yge85[year_cols].T)

              206          430           654
2011      6717973      2755816       9473789
2012      6996707      2919828       9916535
2013            :            :             :
2014            :            :             :
2015            :            :             :
2016            :            :             :
2017            :            :             :
2018            :            :             :
2019            :            :             :
2020            :            :             :
2021  8708349 bep  4288087 bep  12996436 bep
2022   8806127 ep   4365444 ep   13171571 ep
2023  8880717 bep  4434772 bep  13315489 bep
2024   9075199 ep   4604581 ep   13679780 ep
2025   9345128 ep   4822775 ep   14167903 ep


## Missing values
Two distinct patterns identified:
1. Greece (EL): all age categories missing for 2011 only
2. EU27_2020 Y_GE85: missing 2013-2020 - aggregate not calculated by Eurostat until 2021. Individual country data available throughout.

In [18]:
# Check missing values per column
df.isnull().sum() 

freq,unit,sex,age,geo\TIME_PERIOD    0
2011                                 0
2012                                 0
2013                                 0
2014                                 0
2015                                 0
2016                                 0
2017                                 0
2018                                 0
2019                                 0
2020                                 0
2021                                 0
2022                                 0
2023                                 0
2024                                 0
2025                                 0
dtype: int64

There are no missing values. 

In [19]:
# Check if there are any duplicate rows
df.duplicated().sum()

0

There are no duplicate rows.

In [20]:
# Preliminary flag scan across year columns
year_cols = [str(y) for y in range(2011, 2026)]
flagged = df[year_cols].apply(lambda x: x.str.contains('[a-zA-Z]', na=False)).sum().sum()
total = df[year_cols].size
print(f"Flagged values: {flagged} / {total} ({flagged/total*100:.1f}%)")
print(f"Clean values: {total-flagged} / {total} ({(total-flagged)/total*100:.1f}%)")

Flagged values: 767 / 10080 (7.6%)
Clean values: 9313 / 10080 (92.4%)


In [21]:
# Check for Eurostat flags in values
for col in df.columns[1:]:
    flagged = df[col].str.contains('[a-zA-Z]', na=False)
    if flagged.any():
        print(f"{col}: {flagged.sum()} flagged values")

2011: 30 flagged values
2012: 15 flagged values
2013: 1 flagged values
2014: 45 flagged values
2015: 21 flagged values
2016: 21 flagged values
2017: 45 flagged values
2018: 45 flagged values
2019: 24 flagged values
2020: 45 flagged values
2021: 93 flagged values
2022: 72 flagged values
2023: 120 flagged values
2024: 95 flagged values
2025: 95 flagged values


In [22]:
# See what types of flags appear per year
for col in year_cols:
    flags = df[col].str.extract(r'([a-zA-Z]+)')[0].dropna()
    print(f"{col}: {flags.value_counts().to_dict()}")

2011: {'b': 30}
2012: {'b': 15}
2013: {'b': 1}
2014: {'b': 24, 'be': 21}
2015: {'be': 21}
2016: {'be': 18, 'e': 3}
2017: {'b': 24, 'be': 21}
2018: {'p': 24, 'e': 21}
2019: {'b': 24}
2020: {'e': 45}
2021: {'bep': 48, 'b': 24, 'e': 21}
2022: {'ep': 48, 'e': 24}
2023: {'b': 24, 'bep': 24, 'p': 24, 'ep': 24, 'e': 24}
2024: {'ep': 47, 'p': 24, 'e': 24}
2025: {'ep': 47, 'p': 24, 'e': 24}


In [23]:
# Confirm it's the same 24 rows across all flag types
mask = df['2023'].str.contains('[a-zA-Z]', na=False)
print(df[mask][df.columns[0]].value_counts())

freq,unit,sex,age,geo\TIME_PERIOD
A,NR,F,TOTAL,DE_TOT        1
A,NR,F,TOTAL,EU27_2020     1
A,NR,T,Y55-59,PL           1
A,NR,T,Y55-59,FR           1
A,NR,T,Y55-59,EU27_2020    1
                          ..
A,NR,F,Y_GE85,DE_TOT       1
A,NR,F,Y80-84,RO           1
A,NR,F,Y80-84,PL           1
A,NR,F,Y80-84,FR           1
A,NR,T,Y_GE85,RO           1
Name: count, Length: 120, dtype: int64


## Summary

**Dataset:** demo_pjangroup — Population by age group and sex
- **Shape:** 672 rows x 16 columns
- **Period:** 2011-2025
- **Geographies:** 27 EU countries + EU27_2020 aggregate

**Key findings:**
- Combined first column needs splitting into 5 separate columns
- All year columns read as object type due to flags mixed into values
- Leading/trailing whitespace in both column names and values
- Missing values coded as ":" not NaN — pandas doesn't detect automatically
- Greece (EL) missing all 2011 data
- EU27_2020 Y_GE85 missing 2013-2020 — aggregate not available until 2021
- Some values carry Eurostat flags (p=provisional, e=estimated, ep=estimated provisional. Figures are retained as official Eurostat estimates and treated as comparable to final values.

**Actions required in cleaning:**
- Strip whitespace from column names and values
- Split combined first column
- Replace ":" with NaN
- Separate flags from values
- Melt to long format